clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java

```diff
@@ -393,10 +394,14 @@ public class ProducerConfig extends AbstractConfig {
                                         MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION_DOC)
                                 .define(KEY_SERIALIZER_CLASS_CONFIG,
                                         Type.CLASS,
+                                        ConfigDef.NO_DEFAULT_VALUE,
+                                        new ConfigDef.NonNullValidator(),
```

Comment:
I think this is a fair point... people may be using `ProducerConfig`/`ConsumerConfig` instances on their own without using a `KafkaProducer`/`KafkaConsumer` to construct them. In that case, if they haven't specified a value (or have explicitly specified `null`) for the key/value serializer/deserializer class, then upgrading to a version of Kafka that contains this change will cause things to break.

Reply:
1. Both `appendSerializerToConfig` and  `appendDeserializerToConfig` methods are `protected`, and users cannot access them directly.
2. The initialization method of KafkaProducer or KafkaConsumer with ConsumerConfig/ProducerConfig is also `protected`.
3. This validator will only be executed after calling appendDeserializerToConfig/appendSerializerToConfig method .
For example, if the user shows that the value of `key.serializer` is set to null but the valid keySerializer is passed in, the validator will not throw a `ConfigException`.

Reply:
What if someone creates a `ProducerConfig` directly, as opposed to indirectly via a `KafkaProducer`? For example:

```java
Map<String, Object> props = new HashMap<>();
props.put("key.serializer", null);
props.put("value.serializer", null);
props.put("bootstrap.servers", "localhost:9092");
ProducerConfig producerConfig = new ProducerConfig(props);
```

This is fine right now but will break if we add this change.

I'm wondering if people might be constructing `ProducerConfig` and `ConsumerConfig` instances in order to do some kind of validation before storing a configuration somewhere, that's then used later to create a producer/consumer. If so, they might be omitting the (de)serializer properties (or rather, just using `null` as a placeholder) if the (de)serializer isn't known at that time and they plan on using the [constructor](https://github.com/apache/kafka/blob/301c6f44d605f7152f25057c599eeb0b01b2aef9/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L288) that accepts already-instantiated `Serializer`/`Deserializer` instances.

It's not _impossible_ that people are using this pattern but it seems much less likely than some of the other cases that have come up here. It might be alright to keep this. What do you think?

Reply:
Thank you for such a careful review.
I think this problem can be kept, and I hope that adding a validator this time can solve the ConfigException thrown during the initialization of KafkaProducer/KafkaConsumer instead of NullPointerException

Reply:
> I'm unclear on how/why we'd want to change `appendSerializerToConfig`, though?

What I wanted to convey was that if we are modifying the contract to assert that serializer cannot be null at any time in the lifecycle of the `*Config` object, then we need to throw an IllegalArgument exception (in the spirit of fail fast) from `appendSerializerToConfig` if the serializer is null instead of silently ignoring the null input.

Separately, echoing what Chris said above, today, it is [possible](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L272) to create a `ProducerConfig` with key/value serializer set to `null`. After this change, it will not be possible to do so as it will throw an exception. @RivenSun2, can you please help me understand how these use cases with `null` serializers won't be impacted when they upgrade to the version containing this change?

Reply:
Never mind. I posted async and didn't read the conversation above. 
> It might be alright to keep this. What do you think?

I agree as long as we call out this change explicitly in the release notes as a change that could potentially require application code change in certain scenarios.